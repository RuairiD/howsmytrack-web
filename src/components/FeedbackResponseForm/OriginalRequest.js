import React from "react";

import { Col, Row, Typography } from "antd";
import { Div } from "lemon-reset";

import MediaEmbed from "../MediaEmbed/MediaEmbed";

const OriginalRequest = ({ mediaUrl, mediaType, feedbackPrompt }) => (
    <Div>
        <Row gutter={[16, 16]}>
            <Col>
                <MediaEmbed mediaUrl={mediaUrl} mediaType={mediaType} />
            </Col>
        </Row>
        {
            feedbackPrompt && (
                <Row gutter={[16, 16]}>
                    <Col>
                        <Typography.Text strong>The requester has said: </Typography.Text>
                        <Typography.Text>"{feedbackPrompt}"</Typography.Text>
                    </Col>
                </Row>
            )
        }
    </Div>
);

export default OriginalRequest;
