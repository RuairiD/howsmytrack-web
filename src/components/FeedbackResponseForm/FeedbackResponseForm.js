import React, { useState, useEffect } from "react";
import { useMutation } from "react-query";
import ReactGA from "react-ga";
import axios from "axios";

import { Alert, Button, Card, Checkbox, Col, Input, Row } from "antd";
import { Div } from "lemon-reset";
import apiRoot from "../../apiRoot";

import ViewRepliesButton from "../ViewRepliesButton/ViewRepliesButton";
import FeedbackResponseRepliesModal from "../FeedbackResponseRepliesModal/FeedbackResponseRepliesModal";
import OriginalRequest from "./OriginalRequest";

export type FeedbackResponseFormProps = {
    feedbackResponseId: number,
    currentFeedback: string,
    mediaUrl: string,
    mediaType: string,
    feedbackPrompt: string,
    alreadySubmitted: boolean,
    currentAllowReplies: boolean,
    replies: number,
    unreadReplies: number,
};

const SUBMIT_FEEDBACK_RESPONSE_MUTATION = `mutation SubmitFeedbackResponse($feedbackResponseId: Int!, $feedback: String!, $allowReplies: Boolean!) {
    submitFeedbackResponse(feedbackResponseId: $feedbackResponseId, feedback: $feedback, allowReplies: $allowReplies) {
        success
        error
    }
}`;

const GA_FEEDBACK_RESPONSE_CATEGORY = "feedbackResponse";

const FormRow = ({ children }) => (
    <Row gutter={[16, 16]}>
        <Col>
            {children}
        </Col>
    </Row>
);

const FeedbackResponseForm = ({
    feedbackResponseId,
    currentFeedback,
    mediaUrl,
    mediaType,
    feedbackPrompt,
    alreadySubmitted,
    currentAllowReplies,
    replies,
    unreadReplies,
}: FeedbackResponseFormProps) => {
    /*
     * Component for displaying feedback form for a group member to complete.
     */
    const [feedback, setFeedback] = useState(currentFeedback);
    const [allowReplies, setAllowReplies] = useState(currentAllowReplies);
    const [isRepliesModalVisible, setIsRepliesModalVisible] = useState(false);

    useEffect(() => {
        ReactGA.event({
            category: GA_FEEDBACK_RESPONSE_CATEGORY,
            action: "view",
        });
    }, []);

    const onFeedbackTextChange = (event) => {
        setFeedback(event.target.value);
    };

    const onAllowRepliesChange = (event) => {
        setAllowReplies(event.target.checked);
    };

    const submitForm = async () => {
        ReactGA.event({
            category: GA_FEEDBACK_RESPONSE_CATEGORY,
            action: "submit",
        });
        const response = await axios.post(`${apiRoot}/graphql/`, {
            query: SUBMIT_FEEDBACK_RESPONSE_MUTATION,
            variables: {
                feedbackResponseId,
                feedback,
                allowReplies,
            },
        });
        return response.data.data.submitFeedbackResponse;
    };

    const [submitFormMutate, { isLoading, data }] = useMutation(submitForm);

    if (data) {
        if (data.success) {
            ReactGA.event({
                category: GA_FEEDBACK_RESPONSE_CATEGORY,
                action: "success",
            });
        } else {
            ReactGA.event({
                category: GA_FEEDBACK_RESPONSE_CATEGORY,
                action: "error",
            });
        }
    }

    const submitted = alreadySubmitted || (data && data.success);

    const showRepliesModal = () => {
        setIsRepliesModalVisible(true);
    };

    const onRepliesModalCancel = () => {
        setIsRepliesModalVisible(false);
    };

    return (
        <Card>
            <OriginalRequest mediaUrl={mediaUrl} mediaType={mediaType} feedbackPrompt={feedbackPrompt} />
            <Div>
                <FormRow>
                    <Input.TextArea
                        value={feedback}
                        onChange={onFeedbackTextChange}
                        rows={8}
                        disabled={submitted}
                    />
                    {data && data.error && <Alert message={data.error} type="error" showIcon />}
                </FormRow>
                <FormRow>
                    <Checkbox
                        checked={allowReplies}
                        onChange={onAllowRepliesChange}
                        disabled={submitted}
                    >
                        Allow the recipient to anonymously reply to your feedback.
                    </Checkbox>
                </FormRow>
                <FormRow>
                    <Button
                        block
                        type="primary"
                        loading={isLoading}
                        disabled={submitted || !feedback}
                        onClick={submitFormMutate}
                    >
                        {!submitted && "Submit Feedback"}
                        {submitted && "Submitted"}
                    </Button>
                </FormRow>
                {allowReplies && submitted && (
                    <FormRow>
                        <ViewRepliesButton
                            replies={replies}
                            unreadReplies={unreadReplies}
                            onClick={showRepliesModal}
                        />
                    </FormRow>
                )}
            </Div>
            {allowReplies && (
                <FeedbackResponseRepliesModal
                    feedbackResponseId={feedbackResponseId}
                    feedback={feedback}
                    onCancel={onRepliesModalCancel}
                    isVisible={isRepliesModalVisible}
                />
            )}
        </Card>
    );
};

export default FeedbackResponseForm;
