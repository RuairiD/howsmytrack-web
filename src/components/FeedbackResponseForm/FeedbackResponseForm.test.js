import React from "react";
import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import waitForExpect from "wait-for-expect";
import { act } from "react-dom/test-utils";
import axios from "axios";
import FeedbackResponseForm from "./FeedbackResponseForm";

jest.mock("axios");
jest.mock("react-ga");

describe("FeedbackResponseForm", () => {
    it("renders an empty textfield for writing feedback", () => {
        const wrapper = shallow(
            <FeedbackResponseForm
                feedbackResponseId={1901}
                mediaUrl="mediaUrl"
                mediaType="mediaType"
                feedbackPrompt="feedbackPrompt"
            />,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("posts the feedback once it has been written and submitted, disables the form and shows a ViewRepliesButton after", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    submitFeedbackResponse: {
                        success: true,
                        error: null,
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackResponseForm
                feedbackResponseId={1901}
                mediaUrl="mediaUrl"
                mediaType="mediaType"
                feedbackPrompt="feedbackPrompt"
            />,
        );

        act(() => wrapper.find("TextArea").get(0).props.onChange({
            target: {
                value: "feedback",
            },
        }));
        act(() => wrapper.find("Checkbox").get(0).props.onChange({
            target: {
                checked: true,
            },
        }));
        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.find("TextArea").get(0).props.disabled).toBeTruthy();
            expect(wrapper.find("Checkbox").get(0).props.disabled).toBeTruthy();
            expect(wrapper.find("Button").get(0).props.disabled).toBeTruthy();
        });
        expect(wrapper.find("ViewRepliesButton").length).toBe(1);
    });

    it("shows an error if feedback submission fails", async () => {
        axios.post.mockResolvedValue({
            data: {
                data: {
                    submitFeedbackResponse: {
                        success: false,
                        error: "error",
                    },
                },
            },
        });

        const wrapper = mount(
            <FeedbackResponseForm
                feedbackResponseId={1901}
                mediaUrl="mediaUrl"
                mediaType="mediaType"
                feedbackPrompt="feedbackPrompt"
            />,
        );

        act(() => wrapper.find("TextArea").get(0).props.onChange({
            target: {
                value: "feedback",
            },
        }));
        act(() => wrapper.find("Checkbox").get(0).props.onChange({
            target: {
                checked: true,
            },
        }));
        await act(async () => wrapper.find("Button").get(0).props.onClick());

        await waitForExpect(() => {
            wrapper.update();
            expect(wrapper.find("TextArea").get(0).props.disabled).not.toBeTruthy();
            expect(wrapper.find("Checkbox").get(0).props.disabled).not.toBeTruthy();
            expect(wrapper.find("Button").get(0).props.disabled).not.toBeTruthy();
            expect(wrapper.find("Alert").length).toBe(1);
        });
    });

    it("shows a FeedbackResponseRepliesModal if ViewRepliesButton is clicked", () => {
        const wrapper = mount(
            <FeedbackResponseForm
                feedbackResponseId={1901}
                currentFeedback="currentFeedback"
                mediaUrl="mediaUrl"
                mediaType="mediaType"
                feedbackPrompt="feedbackPrompt"
                alreadySubmitted
                currentAllowReplies
                replies={3}
                unreadReplies={2}
            />,
        );

        act(() => wrapper.find("ViewRepliesButton").get(0).props.onClick());
        wrapper.update();

        expect(wrapper.find("FeedbackResponseRepliesModal").get(0).props.isVisible).toBeTruthy();
    });

    it("hides the FeedbackResponseRepliesModal when dismissed", () => {
        const wrapper = mount(
            <FeedbackResponseForm
                feedbackResponseId={1901}
                currentFeedback="currentFeedback"
                mediaUrl="mediaUrl"
                mediaType="mediaType"
                feedbackPrompt="feedbackPrompt"
                alreadySubmitted
                currentAllowReplies
                replies={3}
                unreadReplies={2}
            />,
        );

        act(() => wrapper.find("ViewRepliesButton").get(0).props.onClick());
        wrapper.update();
        act(() => wrapper.find("FeedbackResponseRepliesModal").get(0).props.onCancel());
        wrapper.update();

        expect(wrapper.find("FeedbackResponseRepliesModal").get(0).props.isVisible).not.toBeTruthy();
    });
});
