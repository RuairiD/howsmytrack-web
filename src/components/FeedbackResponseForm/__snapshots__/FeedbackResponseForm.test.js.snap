// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackResponseForm renders an empty textfield for writing feedback 1`] = `
<Card>
  <OriginalRequest
    feedbackPrompt="feedbackPrompt"
    mediaType="mediaType"
    mediaUrl="mediaUrl"
  />
  <Div
    className=""
  >
    <FormRow>
      <TextArea
        onChange={[Function]}
        rows={8}
      />
    </FormRow>
    <FormRow>
      <Checkbox
        indeterminate={false}
        onChange={[Function]}
      >
        Allow the recipient to anonymously reply to your feedback.
      </Checkbox>
    </FormRow>
    <FormRow>
      <Button
        block={true}
        disabled={true}
        ghost={false}
        htmlType="button"
        loading={false}
        onClick={[Function]}
        type="primary"
      >
        Submit Feedback
      </Button>
    </FormRow>
  </Div>
</Card>
`;
